### PR TITLE
feat: add Conditions to Warehouses

### PR DESCRIPTION
**Pull request by hiddeco**
Addresses: #2189

This implements status conditions for Warehouse resources, providing more structured and granular insight into a Warehouse's state. It also enhances the validation of discovered artifacts.

The `Ready` and `Healthy` conditions provide a high-level overview of whether the Warehouse is fully operational and if its current state is normal.

The `Reconciling` condition indicates when a reconciliation is in progress. It is set at the start of reconciliation and removed when reconciliation successfully completes. If reconciliation requires multiple requeues/attempts, the condition remains in place until the reconciliation fully succeeds.

The newly added `DiscoveredAt` field of `DiscoveredArtifacts` enables smarter determination of when to re-discover artifacts based on the user-specified interval (over relying on the requeue value), avoiding unnecessary work (on transient errors).

---

#### Condition logic

- When discovering artifacts:
  - Set `Reconciling` to True with Reason `ScheduledDiscovery`
  - Set `Ready` to False with Reason `AwaitingDiscovery`
  - Set `Healthy` to Unknown with Reason `Pending`
- When artifact discovery fails:
   - Set `Healthy` and `Ready` to False with Reason `DiscoveryFailed`
- When validating discovered artifacts:
  - Set `Healthy` and `Ready` to False if no artifacts found or specific types missing, and remove `Reconciling`
  - Set `Healthy` to True with Reason `ArtifactsDiscovered` if validation passes
- When auto-creating a Freight:
  - Set `Reconciling` to True with Reason `FreightCreationInProgress`
  - Set `Ready` to False with Reason `AwaitingFreightCreation`
- When Freight creation fails:
  - Set `Ready` to False with Reason `FreightCreationFailure`
- After successful reconciliation:
  - Remove the `Reconciling` condition
  - Set `Ready` to True with Reason `ArtifactsDiscovered`

### CODE REVIEW

File: api/v1alpha1/conditions.go [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+	// The meaning of "healthy" is specific to the resource type. For example,
+	// a Warehouse resource may consider itself healthy if it can successfully
+	// connect to its upstream sources, while a Stage resource may consider
+	// itself healthy if the underlying applications are running as expected.
+	//
+	// This is a "normal-true" or "positive polarity" condition, meaning that
+	// the presence of the condition with a status of "True" indicates that
+	// the resource is operating as expected.
+	ConditionTypeHealthy = "Healthy"
+)
```
Comment (krancour) [inline api/v1alpha1/conditions.go:50]: A+ docs. 😍

File: api/v1alpha1/warehouse_types.go [hunk 2]
```diff
@@ -311,13 +311,22 @@ type ChartSubscription struct {
...
 	// LastHandledRefresh holds the value of the most recent AnnotationKeyRefresh
 	// annotation that was handled by the controller. This field can be used to
 	// determine whether the request to refresh the resource has been handled.
 	// +optional
 	LastHandledRefresh string `json:"lastHandledRefresh,omitempty" protobuf:"bytes,6,opt,name=lastHandledRefresh"`
 	// Message describes any errors that are preventing the Warehouse controller
 	// from polling repositories to discover new Freight.
+	//
+	// Deprecated: Use Conditions instead.
 	Message string `json:"message,omitempty" protobuf:"bytes,3,opt,name=message"`
```
Comment (krancour) [inline api/v1alpha1/warehouse_types.go:330]: Can we get away with jumping straight to removing this? I don't event think we would require upgrade logic. Conditions will naturally be filled in in the normal course of business.

File: api/v1alpha1/warehouse_types.go [hunk 3]
```diff
@@ -329,9 +338,21 @@ type WarehouseStatus struct {
...
 }
 
+func (w *WarehouseStatus) GetConditions() []metav1.Condition {
+	return w.Conditions
+}
+
+func (w *WarehouseStatus) SetConditions(conditions []metav1.Condition) {
+	w.Conditions = conditions
+}
+
```
Comment (krancour) [inline api/v1alpha1/warehouse_types.go:348]: Curious about these accessors. Are they here to fulfill some interface?

Comment (krancour) [reply api/v1alpha1/warehouse_types.go:348]: Nevermind. Answered my own question already.

File: api/v1alpha1/warehouse_types.go (hunk 2 above)
Comment (hiddeco) [reply api/v1alpha1/warehouse_types.go:330]: Happy to just remove them. This pathway was mainly to ensure we wouldn't run into backwards compatibility issues, which I think we won't as long as the UI catches up.

**Comment by krancour** [comment]
One very minor question and otherwise LGTM.

To the best of my knowledge, the `warehouse.status.message` field is used in only a couple places in the UI.

I'm wondering if @rbreeze might be able to amend this PR with complementary UI updates, in which case, I suspect we can jump straight to removing the `warehouse.status.message` field as I was inquiring about [here](https://github.com/akuity/kargo/pull/2239#discussion_r1690125372).

File: internal/controller/warehouses/warehouses.go [hunk 4]
```diff
@@ -191,38 +194,169 @@ func (r *reconciler) syncWarehouse(
...
+				Message: fmt.Sprintf(
+					"Discovering artifacts for %d subscriptions",
+					len(warehouse.Spec.Subscriptions),
+				),
+				ObservedGeneration: warehouse.GetGeneration(),
+			},
+			&metav1.Condition{
+				Type:               kargoapi.ConditionTypeReady,
+				Status:             metav1.ConditionFalse,
+				Reason:             "AwaitingDiscovery",
```
Comment (krancour) [inline internal/controller/warehouses/warehouses.go:229]: `AwaitingDiscovery` might be misleading. I think you said this is true when waiting for discovery to complete, but `AwaitingDiscovery` sounds to me like waiting for discovery to _begin_.

Any thoughts on how to clarify?

Comment (hiddeco) [reply internal/controller/warehouses/warehouses.go:229]: `AwaitingDiscoveryResults` or `WaitingForDiscovery` then maybe?

Comment (krancour) [reply internal/controller/warehouses/warehouses.go:229]: `WaitingForDiscovery` feels the same as `AwaitingDiscovery`.

I am fine with `AwaitingDiscoveryResults`... or maybe `DiscoveryInProgress`? Do you feel as if the latter would accurately convey the condition you are trying to convey?

File: api/v1alpha1/warehouse_types.go (hunk 2 above)
Comment (krancour) [reply api/v1alpha1/warehouse_types.go:330]: @rbreeze and I discussed last night and he'll help get the corresponding UI changes into this PR.

**Comment by hiddeco** [comment]
@rbreeze finally got around to testing your UI changes, and it works as expected. Thank you! 🙇 

@krancour this is ready for re-review.

**Comment by krancour** [comment]
@hiddeco I was already happy with the back end of this and if you've reviewed @rbreeze's UI contributions to this, then it's g2g.

**Review by krancour — APPROVED** [review]

File: api/v1alpha1/conditions.go (hunk 1 above)
Comment (Magmax030) [reply api/v1alpha1/conditions.go:50]: Save up

**Pull request merged**